driver-core/src/main/com/mongodb/internal/async/AsyncSupplier.java

```diff
@@ -0,0 +1,137 @@
+     * implemented as a lambda. To "finish" an async chain, use one of
+     * the "finish" methods.
+     *
+     * @see #finish(SingleResultCallback)
+     */
+    void unsafeFinish(SingleResultCallback<T> callback);
+
+    /**
+     * This is the async variant of a supplier's get method.
+     * This method must only be used when this AsyncSupplier corresponds
+     * to a {@link java.util.function.Supplier} (and is therefore being
+     * used within an async chain method lambda).
+     * @param callback the callback
+     */
+    default void getAsync(final SingleResultCallback<T> callback) {
+        unsafeFinish(callback);
+    }
+
+    @Override
+    default void unsafeFinish(final Void value, final SingleResultCallback<T> callback) {
+        unsafeFinish(callback);
+    }
+
+    /**
+     * Must be invoked at end of async chain.
+     * @param callback the callback provided by the method the chain is used in
+     */
+    default void finish(final SingleResultCallback<T> callback) {
+        final boolean[] callbackInvoked = {false};
+        try {
+            this.unsafeFinish((v, e) -> {
+                callbackInvoked[0] = true;
+                callback.onResult(v, e);
+            });
+        } catch (Throwable t) {
+            if (callbackInvoked[0]) {
+                throw t;
+            } else {
+                callback.completeExceptionally(t);
+            }
+        }
+    }
+
+    /**
+     * @param function The async function to run after this supplier
+     * @return the composition of this supplier and the function, a supplier
+     * @param <R> The return type of the resulting supplier
+     */
+    default <R> AsyncSupplier<R> thenApply(final AsyncFunction<T, R> function) {
+        return (c) -> {
+            this.unsafeFinish((v, e) -> {
+                if (e == null) {
+                    function.unsafeFinish(v, c);
+                } else {
+                    c.completeExceptionally(e);
+                }
+            });
+        };
+    }
+
+
+    /**
+     * @param consumer The async consumer to run after this supplier
+     * @return the composition of this supplier and the consumer, a runnable
+     */
+    default AsyncRunnable thenConsume(final AsyncConsumer<T> consumer) {
+        return (c) -> {
+            this.unsafeFinish((v, e) -> {
+                if (e == null) {
+                    consumer.unsafeFinish(v, c);
+                } else {
+                    c.completeExceptionally(e);
+                }
+            });
+        };
+    }
+
+    /**
+     * @param errorCheck A check, comparable to a catch-if/otherwise-rethrow
+     * @param errorFunction   The branch to execute if the error matches
+     * @return The composition of this, and the conditional branch
+     */
+    default AsyncSupplier<T> onErrorIf(
+            final Predicate<Throwable> errorCheck,
+            final AsyncFunction<Throwable, T> errorFunction) {
+        return (callback) -> this.finish((r, e) -> {
```

Comment:
It's surprising that we need `finish` here instead of `unsafeFinish`, given that everywhere else `unsafeFinish` is, as expected, sufficient<sup>1</sup>. Could you leave a comment explaining why `finish` is required here?

---

<sup>1</sup> `AsyncRunnable.thenRunRetryingWhile` also calls `finish`, but there the situation is very different: we must implement `AsyncCallbackSupplier.get` there, and only `finish` meets the contract of `AsyncCallbackSupplier.get`.